airbyte-integrations/connectors/source-amazon-ads/unit_tests/test_streams.py

```diff
@@ -0,0 +1,285 @@
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+from http import HTTPStatus
+from json import dumps, loads
+from urllib.parse import parse_qs, urlparse
+
+import pytest
+import requests
+import responses
+from airbyte_cdk.models import SyncMode
+from jsonschema import validate
+from source_amazon_ads import SourceAmazonAds
+
+
+def setup_responses(
+    profiles_response=None,
+    campaigns_response=None,
+    adgroups_response=None,
+    targeting_response=None,
+    product_ads_response=None,
+    generic_response=None,
+):
+    responses.add(
+        responses.POST,
+        "https://api.amazon.com/auth/o2/token",
+        json={"access_token": "alala", "expires_in": 10},
+    )
+    if profiles_response:
+        responses.add(
+            responses.GET,
+            "https://advertising-api.amazon.com/v2/profiles",
+            body=profiles_response,
+        )
+    if campaigns_response:
+        responses.add(
+            responses.GET,
+            "https://advertising-api.amazon.com/sd/campaigns",
+            body=campaigns_response,
+        )
+    if adgroups_response:
+        responses.add(
+            responses.GET,
+            "https://advertising-api.amazon.com/sd/adGroups",
+            body=adgroups_response,
+        )
+    if targeting_response:
+        responses.add(
+            responses.GET,
+            "https://advertising-api.amazon.com/sd/targets",
+            body=targeting_response,
+        )
+    if product_ads_response:
+        responses.add(
+            responses.GET,
+            "https://advertising-api.amazon.com/sd/productAds",
+            body=product_ads_response,
+        )
+    if generic_response:
+        responses.add(
+            responses.GET,
+            f"https://advertising-api.amazon.com/{generic_response}",
+            json=[],
+        )
+
+
+@responses.activate
+def test_streams_profile(test_config, profiles_response):
+    setup_responses(profiles_response=profiles_response)
+
+    source = SourceAmazonAds()
+    streams = source.streams(test_config)
+    assert len(streams) == 17
+    profile_stream = streams[0]
```

Comment:
So if someone will change the order of the streams then these tests will be hard to update. Why don't we:
1. add new test `test_streams` in `test_source.py` file and move there all code from lines 94-108 (all asserts of stream names)?
Then, it's not convenient to have 17 lines which do the same. We can automate it. We can add a function which will take `stream_name` and `streams` parameters and it will return stream object. For example:
```
def get_stream_by_name(stream_name: str, streams: list):
    for stream in streams:
        if stream.name == stream_name:
            return stream
```
After this we can upgrade `test_streams` function and remove those 17 lines which do the same. So `test_streams` will look something like this:
```
def test_streams(test_config, profiles_response):
    setup_responses(profiles_response=profiles_response)
    source = SourceAmazonAds()
    streams = source.streams(test_config)

    assert len(streams) == 17

    expected_streams_names = ("profiles", "sponsored_display_campaigns", "sponsored_product_campaigns", ...)

    for expected_stream_name in expected_streams_names:
        actual_stream = get_stream_by_name(stream_name=expected_stream_name, streams=streams)
        assert actual_stream is not None
```

Reply:
Fixed